wp-includes/sqlite/class-wp-sqlite-translator.php

```diff
@@ -3488,6 +3441,170 @@ private function execute_show() {
 		}
 	}
 
+	private function generate_create_statement() {
+		$table_name = $this->rewriter->consume()->value;
+		$columns    = $this->get_table_columns( $table_name );
```

Comment:
Thanks for bringing up this point, it made me realize another important detail. I think the two methods are not doing the same. `get_columns_from` is returning a result specifically formatted for the `show columns from` query. However, we could reuse `get_table_columns` inside the `get_columns_from` method.

One problem I now realize I overlooked and further justifies the need for a separate method that returns the raw data is that we need to sort the PK columns. In the `PRAGMA table_info()` query, the position of each PK is returned. The position is not necessarily identical to the order in which they appear. The order of a composite PK is important for indexing and default ordering purposes.

E.g. for the purpose of illustration i created a PK for which the column order is not the same as the order in which the columns are listed. `pragma table_info()` returns:

| cid | name | type | notnull | dflt\_value | pk |
| :--- | :--- | :--- | :--- | :--- | :--- |
| 0 | term\_taxonomy\_id | INTEGER | 1 | 0 | 2 |
| 1 | object\_id | INTEGER | 1 | 0 | 1 |
| 2 | term\_order | INTEGER | 1 | 0 | 0 |


With that, I propose:

1. We use `get_table_columns` inside the `get_columns_from`.
2. I adjust the comment for the `get_columns_from` to clearly indicate its purpose. (which is to return output formatted as `SHOW COLUMNS FROM` MySQL queries).
3. I implement sorting of the PKs so that they appear in the correct order in the final CREATE statement.

What are your thoughts on this?

Reply:
This sounds like a good solution to me, but I'm still wondering if it would be better to update `get_columns_from` to do what you need instead of adding another way of parsing the data from `PRAGMA table_info()`.

Reply:
The way I see it is that we have two distinct desired outputs. One is for `show columns from` and the other `show create table`. Both use the raw data provided by the pragma query under the hood but the goals and outputs are different. `get_columns_from` adds no value over using the raw data directly in this case and i think its best to separate these concerns. Finally, now that we established that we also need to consider the order of the columns in the PK, we can't rely on `get_columns_from` as its output is meant to directly resemble `show columns from` output as MySQL would provide. It does not convey the order of columns in the PK.

This is the output of `get_columns_from`
| Field | Type         | Null | Key | Default | Extra          |
|-------|--------------|------|-----|---------|----------------|
| id    | int(11)      | NO   | PRI | NULL    | auto_increment |
| name  | varchar(255) | YES  |     | NULL    |                |
| date  | date         | YES  |     | NULL    |                |

Do you see another way that this would make sense?

Reply:
> The way I see it is that we have two distinct desired outputs. One is for show columns from and the other show create table

That sounds good. Let's just make their purpose clear with comments.